File: src/binary_file_parser/types/le/int.py
import struct

from binary_file_parser.types.parseable import Parseable
from binary_file_parser.types.byte_stream import ByteStream
from binary_file_parser.types.version import Version


class Int(Parseable):
    __slots__ = "struct_symbol"

    def __init__(self, size: int, struct_symbol: str):
        super().__init__(size)
        self.struct_symbol = struct_symbol

    def _from_stream(self, stream: ByteStream, *, struct_ver: Version = Version((0,))) -> int:
        return self._from_bytes(stream.get(self._size), struct_ver = struct_ver)

    def _from_bytes(self, bytes_: bytes, *, struct_ver: Version = Version((0,))) -> int:
        return struct.unpack(self.struct_symbol, bytes_)[0]

    def _to_bytes(self, value: int) -> bytes:
        return struct.pack(self.struct_symbol, value)

int8 = Int(1, "<b")
int16 = Int(2, "<h")
int32 = Int(4, "<i")
int64 = Int(8, "<q")
uint8 = Int(1, "<B")
uint16 = Int(2, "<H")
uint32 = Int(4, "<I")
uint64 = Int(8, "<Q")

File: src/types/le/int.rs
use pyo3::prelude::*;
use pyo3::pyclass;
use pyo3::types::PyBytes;

use crate::{def_num_type_le, wrap_py};
use crate::types::byte_stream::ByteStream;
use crate::types::parseable::Parseable;
use crate::types::version::Version;


def_num_type_le!(UInt8, "uint8", u8, 1);
def_num_type_le!(UInt16, "uint16", u16, 2);
def_num_type_le!(UInt32, "uint32", u32, 4);
def_num_type_le!(UInt64, "uint64", u64, 8);
def_num_type_le!(UInt128, "uint128", u128, 16);

def_num_type_le!(Int8, "int8", i8, 1);
def_num_type_le!(Int16, "int16", i16, 2);
def_num_type_le!(Int32, "int32", i32, 4);
def_num_type_le!(Int64, "int64", i64, 8);
def_num_type_le!(Int128, "int128", i128, 16);
